### PR TITLE
Fix ModelPruning(make_pruning_permanent=True) buffers getting removed when saved during training

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

During training, every time the model is saved via `ModelCheckpoint`, if `make_pruning_permanent`, the buffers would get removed. This means that the buffers would not accumulate and the pruning amounts are incorrect.

## Before submitting
- [n/a] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/callbacks/pruning.py [hunk 1]
```diff
@@ -381,13 +385,18 @@ def on_train_epoch_end(self, trainer, pl_module, *args):
...
+            rank_zero_debug("`on_train_end`. Pruning permanently...")
+            self.make_pruning_permanent(pl_module)
 
-    def on_save_checkpoint(self, *args):
+    def on_save_checkpoint(self, trainer, pl_module: LightningModule, checkpoint: Dict[str, Any]):
         if self._make_pruning_permanent:
-            self.make_pruning_permanent()
+            rank_zero_debug("`on_save_checkpoint`. Pruning permanently...")
+            # prune a copy so training can continue with the same buffers
+            copy = deepcopy(pl_module)
```
Comment (SeanNaren) [inline pytorch_lightning/callbacks/pruning.py:397]: Just to make sure, does this happen on the CPU mem?

Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py]: Changed!

File: pytorch_lightning/callbacks/pruning.py [hunk 2]
```diff
@@ -381,13 +385,20 @@ def on_train_epoch_end(self, trainer, pl_module, *args):
...
-    def on_save_checkpoint(self, *args):
+    def on_save_checkpoint(self, trainer, pl_module: LightningModule, checkpoint: Dict[str, Any]):
         if self._make_pruning_permanent:
-            self.make_pruning_permanent()
+            rank_zero_debug("`on_save_checkpoint`. Pruning permanently...")
+            prev_device = pl_module.device
+            # prune a copy so training can continue with the same buffers
+            copy = deepcopy(pl_module.to("cpu"))
+            self.make_pruning_permanent(copy)
+            checkpoint["state_dict"] = copy.state_dict()
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:400]: Are Callbacks `on_save_checkpoint` called after `model.state_dict()` in dump_checkpoint ?

File: pytorch_lightning/callbacks/pruning.py [hunk 3]
```diff
@@ -381,13 +385,20 @@ def on_train_epoch_end(self, trainer, pl_module, *args):
...
         if self._make_pruning_permanent:
-            self.make_pruning_permanent()
+            rank_zero_debug("`on_train_end`. Pruning permanently...")
+            self.make_pruning_permanent(pl_module)
 
-    def on_save_checkpoint(self, *args):
+    def on_save_checkpoint(self, trainer, pl_module: LightningModule, checkpoint: Dict[str, Any]):
         if self._make_pruning_permanent:
-            self.make_pruning_permanent()
+            rank_zero_debug("`on_save_checkpoint`. Pruning permanently...")
```
Comment (tchaton) [inline pytorch_lightning/callbacks/pruning.py:395]:
```suggestion
            rank_zero_debug("`ModelPruning.on_save_checkpoint`. Pruning is made permanent for this checkpoint.")
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/callbacks/pruning.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:400]: Yes, see #6072

File: pytorch_lightning/callbacks/pruning.py [hunk 4]
```diff
@@ -381,13 +385,20 @@ def on_train_epoch_end(self, trainer, pl_module, *args):
         ):
             self.apply_lottery_ticket_hypothesis()
 
-    def on_train_end(self, *args):
+    def on_train_end(self, trainer, pl_module: LightningModule):
         if self._make_pruning_permanent:
-            self.make_pruning_permanent()
+            rank_zero_debug("`on_train_end`. Pruning permanently...")
```
Comment (carmocca) [inline pytorch_lightning/callbacks/pruning.py:390]:
```suggestion
            rank_zero_debug("`ModelPruning.on_train_end`. Pruning is made permanent for this checkpoint.")
```

**Review by SkafteNicki — COMMENTED** [review]
smaller comments

**Review by SeanNaren — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]
